dali/imgcodec/util/convert.h

```diff
@@ -68,11 +69,153 @@ void Convert(Out *out, const int64_t *out_strides,
+struct ConvertPixel<Out, In, DALI_GRAY, DALI_GRAY> : ConvertPixelDType<Out, In, 1> {};
+
+template <typename Out, typename In>
+struct ConvertPixel<Out, In, DALI_RGB, DALI_RGB> : ConvertPixelDType<Out, In, 3> {};
+
+template <typename Out, typename In>
+struct ConvertPixel<Out, In, DALI_YCbCr, DALI_YCbCr> : ConvertPixelDType<Out, In, 3> {};
+
+template <typename Out, typename In>
+struct ConvertPixel<Out, In, DALI_GRAY, DALI_RGB> : ColorConversionBase<Out, 1, In, 3> {
+  void operator()(Out *out, const In *in) const {
+    auto rgb = this->vload(in);
+    this->store(out, kernels::color::rgb_to_gray<Out, In>(rgb));
+  }
+};
+
+template <typename Out, typename In>
+struct ConvertPixel<Out, In, DALI_YCbCr, DALI_RGB> : ColorConversionBase<Out, 3, In, 3> {
+  void operator()(Out *out, const In *in) const {
+    auto rgb = this->vload(in);
+    this->vstore(out, kernels::color::itu_r_bt_601::rgb_to_ycbcr<Out, In>(rgb));
+  }
+};
+
+template <typename Out, typename In>
+struct ConvertPixel<Out, In, DALI_YCbCr, DALI_GRAY> : ColorConversionBase<Out, 3, In, 1> {
+  void operator()(Out *out, const In *in) const {
+    auto gray = this->load(in);
+    this->vstore(out, kernels::color::itu_r_bt_601::gray_to_ycbcr<Out, In>(gray));
+  }
+};
+template <typename Out, typename In>
+struct ConvertPixel<Out, In, DALI_RGB, DALI_GRAY> : ColorConversionBase<Out, 3, In, 1> {
+  void operator()(Out *out, const In *in) const {
+    auto gray = ConvertSatNorm<Out>(this->load(in));
+    this->vstore(out, vec<3, Out>{gray, gray, gray});
+  }
+};
+
+template <typename Out, typename In>
+struct ConvertPixel<Out, In, DALI_RGB, DALI_YCbCr> : ColorConversionBase<Out, 3, In, 3> {
+  void operator()(Out *out, const In *in) const {
+    auto ycbcr = this->vload(in);
+    this->vstore(out, kernels::color::itu_r_bt_601::ycbcr_to_rgb<Out, In>(ycbcr));
+  }
+};
+template <typename Out, typename In>
+struct ConvertPixel<Out, In, DALI_GRAY, DALI_YCbCr> : ColorConversionBase<Out, 1, In, 3> {
+  void operator()(Out *out, const In *in) const {
+    auto ycbcr = this->vload(in);
+    this->store(out, kernels::color::itu_r_bt_601::ycbcr_to_gray<Out, In>(ycbcr));
+  }
+};
+
+
+/**
+ * @brief Converts an image stored in `in` and stores it in `out`.
+ *
+ * This is a wrapper for more generic variant of Convert. Based on image format and strides it
+ * chooses an appropriate conversion function and runs the generic Convert with it.
  */
 template <typename Out, typename In>
-inline void ConvertDType(Out *out, const In *in) {
-  *out = ConvertSatNorm<Out>(*in);
+void Convert(Out *out, const int64_t *out_strides, int out_channel_dim, DALIImageType out_format,
+             const In *in, const int64_t *in_strides, int in_channel_dim, DALIImageType in_format,
+             const int64_t *size, int ndim) {
+  // TODO(skarpinski) Support other layouts
+  DALI_ENFORCE(out_channel_dim == ndim - 1 && in_channel_dim == ndim - 1,
+    "Not implemented: currently only channels-last layout is supported");
```

Comment:
any reason we can't support channel first now? I think it's just a matter of adjusting strides, isn't it?

Reply:
And the strides/sizes calculation will be a bit more tricky. Also, high-level convert in https://github.com/NVIDIA/DALI/blob/main/dali/imgcodec/util/convert.cc assumes channel-last in a few places. I think we can leave it as a follow up